src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -39,7 +39,7 @@ internal GeneratorDriver(GeneratorDriverState state)
         internal GeneratorDriver(ParseOptions parseOptions, ImmutableArray<ISourceGenerator> generators, AnalyzerConfigOptionsProvider optionsProvider, ImmutableArray<AdditionalText> additionalTexts, GeneratorDriverOptions driverOptions)
         {
             var incrementalGenerators = GetIncrementalGenerators(generators, SourceExtension);
-            _state = new GeneratorDriverState(parseOptions, optionsProvider, generators, incrementalGenerators, additionalTexts, ImmutableArray.Create(new GeneratorState[generators.Length]), DriverStateTable.Empty, SyntaxStore.Empty, driverOptions, runtime: TimeSpan.Zero, parseOptionsChanged: true);
+            _state = new GeneratorDriverState(parseOptions, optionsProvider, generators, incrementalGenerators, additionalTexts, generators.SelectAsArray(g => GeneratorState.Empty), DriverStateTable.Empty, SyntaxStore.Empty, driverOptions, runtime: TimeSpan.Zero, parseOptionsChanged: true);
```

Comment:
Could you please explain how this change prevents an NRE which was occurring in the original code?

Reply:
I'm guessing it's in `getGeneratorSources` access to `generatorState.PostInitTrees.Length`. The value for `PostInitTrees` is a default immutable array for `default(GeneratorState)`, as opposed to an empty immutable array for `GeneratorState.Empty`.

Reply:
When calling `.GetRunResult()` we call into a local function `getGeneratorSources` that creates an array to handle the generated sources of the given generator

https://github.com/dotnet/roslyn/blob/b8320e1f710db367d33288ed5fee0fb5ee07dd88/src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs#L174

That calls into the collections of the individual states to get their sizes. Prior to the change, we created `default` `GeneratorState`s, so the `PostInitTrees` and `GeneratedTrees` were uninitialized immutable arrays which throw when attempting to get their length.

Instead we now store `GeneratorState.Empty` instances, which in turn have _empty_ result arrays rather than uninitialized ones. 

We had a test that checked you could call `GetRunResult` before calling `RunGenerators` but that test didn't add any generators, so the `generatorStates` array was empty. Adding a generator ensures that array has content, and we see the issue in the added test.